plan/plan.go

```diff
@@ -52,36 +52,47 @@ type Changes struct {
 // processing. It returns a copy of Plan with the changes populated.
 func (p *Plan) Calculate() *Plan {
 	changes := &Changes{}
-
-	// Ensure all desired records exist. For each desired record make sure it's
-	// either created or updated.
-	for _, desired := range p.Desired {
-		// Get the matching current record if it exists.
-		current, exists := recordExists(desired, p.Current)
-
-		// If there's no current record create desired record.
-		if !exists {
-			changes.Create = append(changes.Create, desired)
-			continue
+	usedDesired := map[int]bool{}
+	usedCurrent := map[int]bool{}
+
+	for c, current := range p.Current {
+		for d, desired := range p.Desired {
+			if current.DNSName == desired.DNSName && current.Target == desired.Target &&
+				current.SuitableType() == desired.SuitableType() {
+				usedDesired[d] = true
+				usedCurrent[c] = true
+				log.Debugf("Skipping endpoint %v because target has not changed", desired)
+			}
 		}
+	}
 
-		// If there already is a record update it if it changed.
-		if desired.Target != current.Target {
-			changes.UpdateOld = append(changes.UpdateOld, current)
-
-			desired.RecordType = current.RecordType // inherit the type from the dns provider
-			desired.MergeLabels(current.Labels)     // inherit the labels from the dns provider, including Owner ID
-			changes.UpdateNew = append(changes.UpdateNew, desired)
+	for c, current := range p.Current {
+		if _, found := usedCurrent[c]; found {
 			continue
 		}
-
-		log.Debugf("Skipping endpoint %v because target has not changed", desired)
+		for d, desired := range p.Desired {
```

Comment:
i think the logic here is overcomplicated and needs to be changed, but I leave this to @linki to decide. Essentially what we need to define here (and what is totally transparent here): 
1. What defines record identity - (DNSName + Type) ? What about the dns zone ?
2. What defines record change - only change of the target is supported ?